### ISSUE TITLE
Order of stdout and stderr output?

### ISSUE THREAD

**Issue by mgeier** [jupyter/help#111]
I don't know which exact part(s) of Jupyter cause(s) this, so I'm asking here ...

It's not really a problem for me, but I was wondering why Jupyter seems to change the order of `stdout` and `stderr` outputs.

For example, when I do this in a Jupyter Notebook code cell (`python3 -m notebook`):

```python
import sys

print('stderr1', file=sys.stderr)
print('stdout1')
print('stderr2', file=sys.stderr)
print('stdout2')
```

I get this output:

```
stdout1
stdout2
stderr1
stderr2
```

I would expect the same order as the `print()` calls.

This doesn't happen only in the Notebook, the same thing happens in the Jupyter console (`python3 -m jupyter_console`):

```
In [1]: import sys; print('stderr1', file=sys.stderr); print('stdout1'); print('stderr2', file=sys.stderr); print('stdout2')
stdout1
stdout2
stderr1
stderr2

In [2]: 
```

Strangely, if I start IPython (without mentioning "Jupyter", i.e. `python3 -m IPython`), I get the expected order:

```
In [1]: import sys; print('stderr1', file=sys.stderr); print('stdout1'); print('stderr2', file=sys.stderr); print('stdout2')
stderr1
stdout1
stderr2
stdout2

In [2]: 
```

If I use the plain old Python console (`python3`), I also get the expected order:

```
>>> import sys; print('stderr1', file=sys.stderr); print('stdout1'); print('stderr2', file=sys.stderr); print('stdout2')
stderr1
stdout1
stderr2
stdout2
>>> 
```

What's up with that?

In case that's relevant, I'm doing all this on Debian Linux, using the latest pip-installed releases of the Jupyter and IPython packages, although I might have forgotten to upgrade one or the other of the sub-packages.

**Comment by takluyver** [comment]
There's some code that tries to join outputs on the same stream into one message; I'd guess that's what's doing this. If you print with `flush=True`, is the ordering as you'd expect?

**Comment by mgeier** [comment]
Yes, if I put `flush=True` everywhere, the order is as expected.

But that's not really comforting. In fact, that's totally unexpected: The `stderr` stream is typically unbuffered anyway, so `flush=True` shouldn't change its behavior.
If anything, `stderr` should come before `stdout` (if `flush=False`).
But since the official Python console doesn't do it that way, Jupiter should probably keep in line.

Anyway, is there an actual advantage in arbitrarily interfering with the order?
Why not just pass it on as it comes?

**Comment by takluyver** [comment]
It's not intended to interfere with the order, that's an unintentional side effect of the way the output capture works. I think it's intended to be more efficient when processing many small outputs. It may be possible to add something that flushes stdout when something comes on stderr, and vice versa.

**Comment by mgeier** [comment]
Wasn't that premature optimization in the first place?

If somebody creates hundreds of outputs in a short time, they probably didn't choose the right tool anyway, right?

And if they insist on using it, they might as well wait a few seconds until their output is rendered.

**Comment by takluyver** [comment]
We have had real issues with processing many small outputs, so I wouldn't
call it premature. Yes, if there are thousands and thousands it's probably
a bigger issue, but there's a zone where performance improvements help.

Please focus on solutions; we know the code isn't perfect, and grumbling
about it doesn't make me eager to help.

On 22 Dec 2016 1:16 p.m., "Matthias Geier" <notifications@github.com> wrote:

Wasn't that premature optimization in the first place?

If somebody creates hundreds of outputs in a short time, they probably
didn't choose the right tool anyway, right?

And if they insist on using it, they might as well wait a few seconds until
their output is rendered.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/jupyter/help/issues/111#issuecomment-268797374>, or mute
the thread

**Comment by mgeier** [comment]
I wasn't asking for your help.

On the contrary, I was trying to help you improve Jupyter by pointing out inconsistent behavior.

I would probably also have tried to help with an actual solution, but if you are all bitchy about it, I won't.

**Comment by takluyver** [comment]
Sorry, that was unnecessarily rude. I'm tired. In slightly less than 24 hours, I have another 52 unread messages in my IPython folder, so I don't have much patience for people telling me about 'premature optimisation'. I was trying to explain what's probably happening, why the code works that way, and how it might be improved.

**Comment by minrk** [comment]
Since there's nontrivial overhead on each stream message (*much* more than writing bytes to a terminal/pipe/open file), there is not a 1:1 correlation between `stream.write` and `socket.send`, which allows some coalescing of stdout/stderr messages. Unbuffered output isn't feasible performance-wise, but we could likely trigger flush of stderr on a stdout write and vice versa as @takluyver suggested, which ought to address this case.

**Comment by mgeier** [comment]
So let's do this, shall we?

Which part of Jupiter can this be fixed in?
Where shall I open an issue?

**Comment by takluyver** [comment]
I think it would be in ipykernel. Look at the iostream module, and the method `KernelApp.init_io()`.

**Comment by justinkterry** [comment]
Any progress here?

**Comment by mgeier** [comment]
I've created an issue at the `ipykernel` project: https://github.com/ipython/ipykernel/issues/280

There has been some discussion with @takluyver and @Carreau, but is has not yet been fixed, AFAICT.
